src/Xamarin.Android.Build.Tasks/Tasks/GetMicrosoftNuGetPackagesMap.cs

```diff
@@ -0,0 +1,111 @@
+#nullable enable
+
+using System;
+using System.IO;
+using Microsoft.Build.Framework;
+using Microsoft.Android.Build.Tasks;
+using System.Net.Http;
+using System.Diagnostics.CodeAnalysis;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Xamarin.Android.Tasks;
+
+public class GetMicrosoftNuGetPackagesMap : AndroidAsyncTask
+{
+	public override string TaskPrefix => "GNP";
+
+	/// <summary>
+	/// The cache directory to use for Maven artifacts.
+	/// </summary>
+	[Required]
+	public string MavenCacheDirectory { get; set; } = null!; // NRT enforced by [Required]
+
+	[Output]
+	public string? ResolvedPackageMap { get; set; }
+
+	public override async System.Threading.Tasks.Task RunTaskAsync ()
+	{
+		Directory.CreateDirectory (MavenCacheDirectory);
+
+		// We're going to store the resolved package map in the cache directory as
+		// "microsoft-packages-{YYYYMMDD}.json". If the file is older than today,
+		// we'll try to download a new one.
+		var all_files = PackagesFile.FindAll (MavenCacheDirectory);
+
+		if (!all_files.Any (x => x.IsToday)) {
+			// No file for today, download a new one
+			try {
+				var http = new HttpClient ();
+				var json = await http.GetStringAsync ("https://aka.ms/ms-nuget-packages");
+				var outfile = Path.Combine (MavenCacheDirectory, $"microsoft-packages-{DateTime.Today:yyyyMMdd}.json");
```

Comment:
Should we actually put this file in `obj`, so if users do a `Clean` it will download again?

Reply:
It feels like this file is similar to a NuGet artifact and should be shared and cached across projects.  If you are working without internet, it would also be nice if you got to keep using one you previously downloaded.

This logic will attempt to download a new one each day, but will keep the previous one if it can't get a new one.

I don't think this file will change extremely frequently as it does not include version numbers.  It will only change when we bind a new AndroidX/GPS library.

Reply:
I'm concerned about this "per-day" file:

The filesystem is a machine-global, barely concurrent-friendly, "database" with which it's easy to cause "spooky action at a distance".

`MavenCacheDirectory` is `$(MavenCacheDirectory)`, which is e.g. `$(LocalAppData)\dotnet-android\MavenCacheDirectory\`.

What happens if, through some terrible confluence of events, *two* projects are built at the same time, on the same machine, for the first time "today"?  Depending on timing, they may each see that today's `microsoft-packages-*.json` doesn't exist, each attempt to *create* it, and *one* of them will "lose", possibly causing some variation on our favorite "file share exception" or whatever.

This feels like a potential source of build *un*-reliability.

This also feels like an impediment to future changes: we can't change the file format.  If we have both .NET 9 and .NET 10 installed on a given machine, and these files are stored "machine global", then there is an implicit requirement that .NET 9 be able to consume .NET 10 output, i.e. the file format either can't change, or the file format needs to be "extensible".

I'd feel much "safer" if this file *were* stored in `$(IntermediateOutputPath)`.  Lower risk of file share -related errors due to concurrent project builds, no chance of file format-related mismatches.

Reply:
> What happens if, through some terrible confluence of events, two projects are built at the same time, on the same machine, for the first time "today"? Depending on timing, they may each see that today's microsoft-packages-*.json doesn't exist, each attempt to create it, and one of them will "lose", possibly causing some variation on our favorite "file share exception" or whatever.

If this scenario were to happen, the "losing" project would log a build message saying it couldn't download the file and any dependency verification errors would not provide NuGet recommendations. On subsequent builds the NuGet recommendations would be available.

This file is considered "optional" and simply enhances the error messages.  However failure is a possible outcome (like if the user doesn't have internet) that is handled and still provides the core experience.

> i.e. the file format either can't change, or the file format needs to be "extensible"

The file format is JSON, making it easily extensible if desired in the future.  If it needed to be extended somehow that JSON couldn't handle we would likely just rename the new version to something like `microsoft-packages-v2.json`.

> I'd feel much "safer" if this file were stored in $(IntermediateOutputPath). Lower risk of file share -related errors due to concurrent project builds, no chance of file format-related mismatches.

I really like that we're only downloading this file once per day no matter how many projects one is using, and that any project can use the machine-wide one if internet isn't available.  But we can change it if we still have strong concerns.